chore(scripts): add base release initializer

mujahidkay opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

<!-- Integration testing generally doesn't run until a PR is labeled for merge, but can be opted into for every push by adding label 'force:integration', and can be customized to use non-default external targets by including lines here that **start** with leading-`#` directives:
* (https://github.com/Agoric/documentation) #documentation-branch: $BRANCH_NAME
* (https://github.com/endojs/endo) #endo-branch: $BRANCH_NAME
* (https://github.com/Agoric/dapp-offer-up) #getting-started-branch: $BRANCH_NAME
* (https://github.com/Agoric/testnet-load-generator) #loadgen-branch: $BRANCH_NAME

These directives should be removed before adding a merge label, so final integration tests run against default targets.
-->

<!-- Most PRs should close a specific issue. All PRs should at least reference one or more issues. Edit and/or delete the following lines as appropriate (note: you don't need both `refs` and `closes` for the same one): -->

closes: #11142

## Description
<!-- Add a description of the changes that this PR introduces and the files that are the most critical to review. -->
Adds an `init-release.sh` script to automate the mundane, repetitive parts of release initialization. Note that this only makes the necessary changes and other need-based changes go case-by-case (and manually) (for example, release with variants). 

Currently, it does the following tasks:
 - updates handler name in `app.go`
 - updates handler name, upgrade names in `upgrade.go`
 - renames `n:upgrade-next` to `a:upgrade-<release-number>`
 - updates the README in `a:upgrade-<release-number>`

In the future, I will keep automating other bits and pieces of release process. However, next on my docket is to update this script to incorporate variants (at least until we get rid of them altogether).

### Security Considerations
<!-- Does this change introduce new assumptions or dependencies that, if violated, could introduce security vulnerabilities? How does this PR change the boundaries between mutually-suspicious components? What new authorities are introduced by this change, perhaps by new API calls? -->
None

### Scaling Considerations
<!-- Does this change require or encourage significant increase in consumption of CPU cycles, RAM, on-chain storage, message exchanges, or other scarce resources? If so, can that be prevented or mitigated? -->
None

### Documentation Considerations
<!-- Give our docs folks some hints about what needs to be described to downstream users.  Backwards compatibility: what happens to existing data or deployments when this code is shipped? Do we need to instruct users to do something to upgrade their saved data? If there is no upgrade path possible, how bad will that be for users? -->
~~Will add to MAINTAINERS.md in a follow up.~~
Updated MAINTAINERS.md 

### Testing Considerations
<!-- Every PR should of course come with tests of its own functionality. What additional tests are still needed beyond those unit tests? How does this affect CI, other test automation, or the testnet? -->
None for the sdk codebase. But for PR verification, can run it locally on master using `./init-release.sh 19` and it will do the needful.

### Upgrade Considerations
<!-- What aspects of this PR are relevant to upgrading live production systems, and how should they be addressed? What steps should be followed to verify that its changes have been included in a release (ollinet/emerynet/mainnet/etc.) and work successfully there? If the process is elaborate, consider adding a script to scripts/verification/. -->
None, this is a helper script for release process.

mujahidkay commented:
@mergifyio rebase

mhofman commented:
My bash script-fu is low. I think @gibson042 might be a better reviewer.

gibson042 commented on scripts/init-release.sh:
```diff
@@ -0,0 +1,79 @@
+  fi
+}
+
+validate_args "$@"
+update_app_go
+update_upgrade_go_handler
+update_upgrade_names
+update_primary_upgrade_name
+rename_upgrade_next_dir
```
I don't think the functions are actually carrying their weight. Suggestion: switch to a single-level imperative script.
```sh
#!/bin/bash
set -ueo pipefail

if [ "$#" -ne 1 ]; then
  echo "Usage: $0 <next release number>"
  echo "Example: $0 19"
  exit 1
fi

# Silly old Darwin.
BSD_SED="$(sed --help 2>&1 | sed 2q | grep -qe '-i ' && echo 1 || true)"
function sedi () {
  if [ -n "$BSD_SED" ]; then
    sed -i '' "$@"
  else
    sed -i "$@"
  fi
}

# Validate arguments.
NEXT_RELEASE=$1
if ! [[ "$NEXT_RELEASE" =~ ^[1-9][0-9]*$ ]]; then
  echo "Error: Release number must be a positive decimal integer"
  exit 1
fi

# Rename function `unreleasedUpgradeHandler`.
sedi "
  s/unreleasedUpgradeHandler/upgrade${NEXT_RELEASE}Handler/g;
  s/the unreleased upgrade/upgrade-${NEXT_RELEASE}/g;
" \
  golang/cosmos/app/app.go \
  golang/cosmos/app/upgrade.go
echo "Renamed unreleasedUpgradeHandler to `upgrade${NEXT_RELEASE}Handler`"

# Update cosmos-sdk upgrade names.
sedi '/^var upgradeNamesOfThisVersion = \[\]string{/,/^}/ c\var upgradeNamesOfThisVersion = []string{\
    "agoric-upgrade-'"${NEXT_RELEASE}"'",\
}' golang/cosmos/app/upgrade.go
echo "Updated cosmos-sdk upgrade names"

…
```

gibson042 reviewed (changes requested):
Overall, I think the effort makes sense. Please also update [MAINTAINERS.md](https://github.com/Agoric/agoric-sdk/blob/master/MAINTAINERS.md) as appropriate.

mujahidkay commented:
> Overall, I think the effort makes sense. Please also update [MAINTAINERS.md](https://github.com/Agoric/agoric-sdk/blob/master/MAINTAINERS.md) as appropriate.

Done. I have incorporated your suggested and updated MAINTAINERS.md. PTAL at your convenience.

mujahidkay commented:
@Mergifyio rebase

mujahidkay commented:
@mergifyio requeue

mujahidkay commented:
@mergifyio rebase

mujahidkay commented:
@mergifyio refresh
